xarray/coding/times.py

```diff
@@ -700,12 +704,26 @@ def _cast_to_dtype_if_safe(num: np.ndarray, dtype: np.dtype) -> np.ndarray:
     return cast_num
 
 
+@overload
+def encode_cf_datetime(
+    dates: chunkedduckarray,
+    units: str | None = None,
+    calendar: str | None = None,
+    dtype: np.dtype | None = None,
+) -> tuple[chunkedduckarray, str, str]: ...
+@overload
+def encode_cf_datetime(
+    dates: duckarray,
+    units: str | None = None,
+    calendar: str | None = None,
+    dtype: np.dtype | None = None,
+) -> tuple[duckarray, str, str]: ...
 def encode_cf_datetime(
-    dates: T_DuckArray,  # type: ignore
+    dates: duckarray | chunkedduckarray,
```

Comment:
You should still use TypeVars and not the protocol directly.

Reply:
I don't think that works (yet, waiting on Higher-Kinded TypeVars). Using `T_DuckArray` will indeed retain the `np.ndarray` typing (good) but also the `dtype` and `shape` typing (bad/wrong). 

We don't really care (in theory) if it is ndarray or cupy array, just that they pass the minimal requirements for a `duckarray`. But we do care that `dtype` and `shape` is correct after calculation.

`duckarray` is actually `duckarray[_ShapeType, _DType]` where the dtype and shape are TypeVars, in same fashion as numpy see #8294. I am intentionally lazy with this in non-namedarray files though to avoid the PRs touching every file.